sdk/servicebus/Microsoft.Azure.ServiceBus/tests/SenderReceiverClientTestBase.cs

```diff
@@ -275,6 +275,103 @@ internal async Task CancelScheduledMessagesAsyncTestCase(IMessageSender messageS
             Assert.True(count == messageCount);
         }
 
+        internal async Task MessageHandlerUnregisterAsyncTestCase(
+            IMessageSender messageSender,
+            IMessageReceiver messageReceiver,
+            int maxConcurrentCalls,
+            bool autoComplete,
+            int messageCount)
+                {
+                    var count = 0;
+                    await TestUtility.SendMessagesAsync(messageSender, messageCount);
+                    messageReceiver.RegisterMessageHandler(
+                        async (message, token) =>
+                        {
+                            TestUtility.Log($"Received message: SequenceNumber: {message.SystemProperties.SequenceNumber}");
+
+                            // Method should be thread safe and stop the message handler after first batch of messages
+                            messageReceiver.UnregisterMessageHandler();
+
+                            // Simulate "long running" task
+                            Thread.Sleep(1000);
+                                                       
+                            if (messageReceiver.ReceiveMode == ReceiveMode.PeekLock && !autoComplete)
+                            {
+                                // Completion should still work, even after message handler is stopped
+                                await messageReceiver.CompleteAsync(message.SystemProperties.LockToken);
+                            }
+
+                            Interlocked.Increment(ref count);
+                        },
+                        new MessageHandlerOptions(ExceptionReceivedHandler) { MaxConcurrentCalls = maxConcurrentCalls, AutoComplete = autoComplete });
+
+                    // Wait for the OnMessage Tasks to finish
+                    var stopwatch = Stopwatch.StartNew();
+                    while (stopwatch.Elapsed.TotalSeconds <= 30)
```

Comment:
>30 [](start = 61, length = 2)

4 different permutations of this test, which essentially means 2 mins added run time. Not a blocker, but maybe consider optimizing this to a smaller number.